### ISSUE TITLE
[CI] Meta-Issue for Gitlab migration

### ISSUE THREAD

**Issue by ejgallego** [coq/coq#6919]
It is not a secret that Coq CI demands are pretty high. As of today we need more than 10 hours build time of a medium-size CPU instance to test one PR. Would we do more serious testing this constraint would easily multiply by 10, so free plans are not an option. Custom commercial plans could be considered, however, Gitlab offers the possibility of running our own runners in our own servers. This is very attractive for several reasons:

- No time limit per build: we can test developments that take a few hours to compile.
- No limit on the number of instances: Gitlab can spin spot AWS instances so latency is minimal.
- Open source: Gitlab is the only mature integral hosting/CI infrastructure that is open source.
- Cleanup of the CI system: right now we support 3 different systems and we cannot standardize on a single platform, with the overhead this implies.
- All kind of custom artifact building and deployment.

Most of the work to migrate to Gitlab is already in place thanks to @SkySkimmer . I have setup a couple of reasonably powerful runners on MINES test servers so we should be covered in this sense too. The runners have been tested and they do seem to work and don't involve any undesired hack. A possible plan of action is:

- s̶e̶t̶u̶p̶ ̶a̶ ̶p̶r̶o̶p̶e̶r̶ ̶C̶o̶q̶ ̶m̶i̶r̶r̶o̶r̶ ̶a̶t̶ ̶h̶t̶t̶p̶s̶:̶/̶/̶g̶i̶t̶l̶a̶b̶.̶c̶o̶m̶/̶c̶o̶q̶/̶c̶o̶q̶ ̶,̶ ̶r̶e̶g̶i̶s̶t̶e̶r̶ ̶t̶h̶e̶ ̶`̶e̶t̶i̶l̶i̶o̶`̶ ̶r̶u̶n̶n̶e̶r̶ ̶[̶3̶2̶ ̶c̶o̶r̶e̶s̶/̶3̶2̶G̶i̶B̶ ̶R̶A̶M̶]̶.̶
- m̶i̶s̶c̶e̶l̶l̶a̶n̶e̶a̶ ̶o̶p̶t̶i̶m̶i̶z̶a̶t̶i̶o̶n̶.̶ ̶F̶o̶r̶ ̶e̶x̶a̶m̶p̶l̶e̶ ̶w̶e̶ ̶m̶a̶y̶ ̶w̶a̶n̶t̶ ̶t̶o̶ ̶a̶d̶d̶ ̶t̶a̶g̶s̶ ̶t̶o̶ ̶a̶v̶o̶i̶d̶ ̶t̶h̶e̶ ̶s̶h̶a̶r̶e̶d̶ ̶r̶u̶n̶n̶e̶r̶s̶ ̶i̶n̶ ̶C̶P̶U̶-̶i̶n̶t̶e̶n̶s̶i̶v̶e̶ ̶j̶o̶b̶s̶,̶ ̶o̶r̶ ̶t̶o̶ ̶p̶u̶l̶l̶ ̶s̶p̶e̶c̶i̶f̶i̶c̶,̶ ̶f̶a̶s̶t̶e̶r̶ ̶i̶m̶a̶g̶e̶s̶ ̶f̶o̶r̶ ̶t̶h̶e̶ ̶c̶o̶m̶m̶o̶n̶ ̶b̶u̶i̶l̶d̶s̶.̶ #6962 
-̶a̶d̶d̶ ̶n̶o̶t̶i̶f̶i̶c̶a̶t̶i̶o̶n̶s̶ ̶t̶o̶ ̶G̶i̶t̶h̶u̶b̶ ̶i̶n̶t̶e̶r̶f̶a̶c̶e̶.̶ ̶G̶i̶t̶l̶a̶b̶ ̶m̶a̶s̶t̶e̶r̶ ̶h̶a̶s̶ ̶[̶a̶l̶r̶e̶a̶d̶y̶ ̶s̶u̶p̶p̶o̶r̶t̶ ̶f̶o̶r̶ ̶t̶h̶i̶s̶]̶(̶ ̶ ̶h̶t̶t̶p̶s̶:̶/̶/̶g̶i̶t̶l̶a̶b̶.̶c̶o̶m̶/̶g̶i̶t̶l̶a̶b̶-̶o̶r̶g̶/̶g̶i̶t̶l̶a̶b̶-̶e̶e̶/̶b̶l̶o̶b̶/̶m̶a̶s̶t̶e̶r̶/̶d̶o̶c̶/̶u̶s̶e̶r̶/̶p̶r̶o̶j̶e̶c̶t̶/̶i̶n̶t̶e̶g̶r̶a̶t̶i̶o̶n̶s̶/̶g̶i̶t̶h̶u̶b̶.̶m̶d̶)̶
- setup AWS on spot instances so extra capacity is there when needed.
- find a new server / fully more to AWS.

This should solve the following issues #6347 #6229 #5602. Issue #6807 needs attention after completing this.

**Comment by ejgallego** [comment]
Need help with points 1 and 2; I can take care of 3 and 4 [provided INRIA opens an AWS account]

**Comment by SkySkimmer** [comment]
>add notifications to Github interface.

What does this mean?

**Comment by ejgallego** [comment]
> What does this mean?

It does mean that Gitlab CI status would ideally be shown on Github's PRs.

**Comment by ejgallego** [comment]
With regards to mirroring, it seems that the problem is that the Coq repository is not public, so this would be only enabled for a pay plan. So here we either use a bot to push to the private repos, or make the clone public.

**Comment by ejgallego** [comment]
More on GH/GL integration: https://gitlab.com/gitlab-org/gitlab-ce/issues/32052

**Comment by Zimmi48** [comment]
I think a mirror is not what we need anyways. We need a repository on which a bot (force-)pushes to branches corresponding to PRs opened on the GitHub repository. I can try to look into this and also the GitHub status updates.

I said it orally before but I wouldn't mind BTW if this repository was deleted with all the duplicate issues it contains then recreated but not as a mirror. I think you said I was an admin on the GitLab Coq organization. Should I go ahead and do it?

**Comment by ejgallego** [comment]
Sure @Zimmi48 you have total freedom to do what you want. Note that you can disable the issues in the repository setting and that could have the same effect. Mirroring could be very effective but it is only available to public repos as it is otherwise a pay feature.

Once you think things are OK you need to generate a token from Settings -> CI and send it to me in private, then my runner will start picking up jobs.

**Comment by Zimmi48** [comment]
[This comment](https://gitlab.com/gitlab-org/gitlab-ee/issues/3836#note_56949346) should provide a good basis for status checks.

**Comment by SkySkimmer** [comment]
If we make the repo private doesn't that mean only those in the gitlab organisation can look at CI results? That sounds undesirable for external contributors, and it means we have to maintain the org membership just for visibility.

**Comment by ejgallego** [comment]
Good point. I was thinking that the PR update hook could push the link, but indeed that link will remain private, right?

**Comment by SkySkimmer** [comment]
Yes.
It might be possible to make only CI visible but why bother? Also I'm 
not sure it's possible.

Gaëtan Gilbert

**Comment by SkySkimmer** [comment]
OK I don't think it's possible, the visibility option for sub items has 
"everyone with access" as the most general setting so the project would 
need to be public.

Gaëtan Gilbert

**Comment by gares** [comment]
I fully acknowledge the problem.

I'm a bit rusty. Isn't AWS a paying service (as circle/travis is?)

I thought that the main advantage of travis/circle/appveyor/gitlab was that we are not the admins of the platform (i.e. someone else does the sysadmin work).

**Comment by ejgallego** [comment]
> I'm a bit rusty. Isn't AWS a paying service (as circle/travis is?)

Yup, but a fraction of the price of any other alternative. [Unless INRIA wants to provide runners]

> I thought that the main advantage of travis/circle/appveyor/gitlab was that we are not the admins of the platform (i.e. someone else does the sysadmin work).

Yup, however they have many other disadvantages such as not working at all.

Sysadmin work with on demand instances is fairly fairly low. [just spin a prebuilt image, then you kill it at the end]

**Comment by Zimmi48** [comment]
I agree with @SkySkimmer that the repository will have to be public, but that shouldn't be a problem if the repository does not contain duplicates of issues and pull requests, and is clearly marked "for CI".

**Comment by ejgallego** [comment]
Looks like a good plan; however you may need to push the PR as to request triggers?

I think you can block however users from messing with public repositories thou. Something I didn't try is the mirroring feature that I understand has advanced a lot [and can be set so issues at least are ignored]

I did the mistake of creating the private repo and then it is disabled. Maybe worth having a look,  the lock on `gitlab/coq` is all yours.

**Comment by SkySkimmer** [comment]
Gitlab released their github integration and I tried using it. It worked on a test repo (see https://github.com/SkySkimmer/gitlab-test/pull/1). It worked on a Coq PR from me to myself (https://github.com/SkySkimmer/coq/pull/9). 
However nothing is happening so far on a test PR from @ejgallego's repo to mine (https://github.com/SkySkimmer/coq/pull/10).

**Comment by Zimmi48** [comment]
I anticipated the problem: mirroring does not work for PRs only for branches. Furthermore, what happens when you force-push to your topic branch. Does it rebuild anything? In any case, we would need a bot to (force-)push PRs to new branches on GitLab. The integration will at least save us from handling the reverse direction (status checks).

**Comment by Zimmi48** [comment]
BTW @ejgallego said it would also solve #6229. Can we have several status checks going back from GitLab CI to GitHub like we have right now with CircleCI?

**Comment by SkySkimmer** [comment]
>I can't find the original repository so hard to know that question.

WDYM? What original repo? What question?

>BTW @ejgallego said it would also solve #6229. Can we have several status checks going back from GitLab CI to GitHub like we have right now with CircleCI?

I think they've got an issue for per-stage statuses but that's not great for us.

**Comment by ejgallego** [comment]
> WDYM? What original repo? What question?

https://gitlab.com/SkySkimmer/gitlab-test/pipelines/19371865 is dead.

> I think they've got an issue for per-stage statuses but that's not great for us.

Yeah, so we need to request per-job status.

**Comment by SkySkimmer** [comment]
Oh for some reason the repo was private, I guess that's the default. It should be accessible now.

**Comment by ejgallego** [comment]
I see, thanks. So the pipeline has only one job then. I guess it does the same for many jobs as you've said.

Question, can't you run stages in parallel?

Anyways, we can start with this simple integration, and keep progressing forward. We are still missing more runners etc... so this is not something we will have ready in one day.

**Comment by maximedenes** [comment]
I'm happy to help with it. I can provide 20 virtual machines as workers, and 2 physical machines. I can also help with the integration. Will create a github project now so we can dispatch the various tasks.

**Comment by maximedenes** [comment]
I deleted and re-created the gitlab coq/coq repo, as a CI/CD-only repo. I'm setting up (signed) Windows and macOS package generation for branches. I activated the GitHub integration.

The only missing bit to complete the Gitlab CI infrastructure is a bot that pushes branches from PRs to another Gitlab fork, that performs untrusted builds. @Zimmi48 said he would have a look at it this week.

**Comment by ejgallego** [comment]
Cool, thanks @maximedenes ; another thing we need is a dedicated build server, I have an idea on what the right hardware should be [Coq CI is really CPU bound so a Xeon with less, but faster cores will work best].

The setup is very very easy tho load-tuning is delicate as if you overload the server docker tends to get stuck; the dedicated setup I had was pretty reliable, the one with the shared server was not due to other users bringing the server down to its knees.

**Comment by Zimmi48** [comment]
As promised I had a look and it was not too difficult. I currently have a "bot" (server pinged by a GitHub webhook) running on my computer and pushing automatically to GitLab, then GitLab does the rest of the job (showing the PR status, first as pending, then as success or failure). Of course that's just a test: the next step will be to deploy this on an actual server.

**Comment by ejgallego** [comment]
That's very cool, merci Théo; I wonder how the server setup is going. I found that doing a stable setup is not very difficult with Ubuntu + Docker + gitlab-runner, but docker however seems sensitive to load; there are several remedies to that, but just a bit conservative resources limits can do the job IMHO.

**Comment by Zimmi48** [comment]
The synchronization bot is now using a dedicated @coqbot account at GitLab instead of my personal account. On the contrary, synchronization of the master branch is always marked as "@maximedenes pushed". Not sure why, but we might want to use @coqbot there as well...

**Comment by Zimmi48** [comment]
It seemed that there were unprotected secret variables on GitLab CI (for testing the Windows packaging stuff IIUC). Given that the synchronization of PRs with GitLab is currently on, I've protected them again.

**Comment by ejgallego** [comment]
We have done a lot of progress in this area, including #7387 and #6962 .

IMHO we are very close to being able to rely on Gitlab for all testing purposes. I will write an update on the plan later this week.

**Comment by ejgallego** [comment]
@maximedenes please give an update on your side when you can so we can coordinate.

**Comment by maximedenes** [comment]
Current status on my side:
- I have a cloudstack VM that can build and sign Windows packages from Gitlab CI.
- I started working on something similar for macOS

My todo-list (work planned this week):
- Finish the macOS runner setup
- Move the trusted runners (which can sign) to my fork and use the "deployment" feature to make signing a one-click operation (as opposed to the current zero-click one)
- Put two other runners on the main Gitlab repo, to build unsigned Windows and macOS packages

Once that is done, I'll use the remaining cloudstack resources to provide Linux runners for the main repo.

**Comment by ejgallego** [comment]
So what kind of setup do the Linux cloudstack runners need?

**Comment by ejgallego** [comment]
> I have a cloudstack VM that can build and sign Windows packages from Gitlab CI.

Will this replace appveyor?

> I started working on something similar for macOS

Will this replace Travis?

Interesting you chose to work on the OSX / Windows subsystem first, I had thought that we were doing pretty well on that front with AppVeyor / Travis and the most pressing problem was indeed our shortage of CI Linux-based runners, why this order of action?

**Comment by ejgallego** [comment]
> So what kind of setup do the Linux cloudstack runners need?

Some info here: https://wiki.inria.fr/ciportal/CloudStack_Tutorial

**Comment by maximedenes** [comment]
> Interesting you chose to work on the OSX / Windows subsystem first, I had thought that we were doing pretty well on that front with AppVeyor / Travis and the most pressing problem was indeed our shortage of CI Linux-based runners, why this order of action?

Having to setup VMs and sign manually was putting a lot of pressure on the final steps of the release. It basically makes me work outside work hours every two months or so (major and minor releases), and puts some pressure on my family life, which is a priority for me.

> So what kind of setup do the Linux cloudstack runners need?

I don't know exactly, but I think we can either use ready-made images or contribute some, based on whatever distro we like.

> Will this replace appveyor?

I think we can keep appveyor for the OPAM-based test-suite part, in order to save some cloudstack resources. But the appveyor build job can be removed.

> Will this replace Travis?

Same.

**Comment by ejgallego** [comment]
> Having to setup VMs and sign manually was putting a lot of pressure on the final steps of the release.

I had understood that the image generation for releases was automatic, it seems it is not the case, I see. Any chance we can make the release documents public? [And kill that private repos for once and for all]

**Comment by maximedenes** [comment]
There is no up-to-date release documents, so we need to write them down. #6057 is what we have at the moment.

**Comment by ejgallego** [comment]
> #6057 is what we have at the moment.

#6057 is cool but yeah, the final steps seem to be missing.

**Comment by Zimmi48** [comment]
Most is written down already. This PR is waiting on feedback for the moment.

**Comment by Zimmi48** [comment]
Indeed, see what I wrote in #6229.

To me what remains to do for the migration is:
- [ ] Ensure that GitLab CI is reliable enough for CI testing.
- [ ] Backport the sharing improvements done in CircleCI to GitLab CI.
- [ ] Remove CircleCI (and possibly Travis as well, if everything in there is on GitLab as well).
- [ ] Document all this.

**Comment by ejgallego** [comment]
> Backport the sharing improvements done in CircleCI to GitLab CI.

Sorry Théo (@Zimmi48 ), what do you mean by this one?

**Comment by Zimmi48** [comment]
@ejgallego Look at the difference of topology between these two:

![circleci-pipeline](https://user-images.githubusercontent.com/1108325/39699787-edba0f12-51fa-11e8-916b-dfdc102356ca.png)

![gitlab-pipeline](https://user-images.githubusercontent.com/1108325/39699741-bfd57104-51fa-11e8-8cd0-59c9b766516d.png)

**Comment by ejgallego** [comment]
Ah, I know what you mean now. I think this is not easy to do in Gitlab now as stages cannot really do this well. I ignore if Gitlab has true dependency-based building on their roadmap or not.

**Comment by SkySkimmer** [comment]
The only real improvement is sharing bignums between color and the mathclasses/corn/formal topology line. bignums isn't very big so it doesn't improve much, I think the loss of parallelism from separating jobs into stages would make it go slower on gitlab.
If it's just a chain without sharing one product with multiple parallel users we may as well gather them in one job.

**Comment by ejgallego** [comment]
I do agree with @SkySkimmer 's analysis of the current situation. But maybe we'd like to keep Circle for precisely this kind of setup, if we find out that the current job packing in Gitlab limits us in the future. I dunno, but today indeed we don't lose a lot.

**Comment by SkySkimmer** [comment]
I think these days circle completes after gitlab though.

Gaëtan Gilbert

**Comment by ejgallego** [comment]
> I think these days circle completes after gitlab though.

Would we remove most redundant jobs situation would improve tho. In fact we could start doing that soon, even if it'd be great if we could improve the display on GitHub first so it is per-job.

Circle would only make sense for the kind of topology "very expensive job followed by a few quick dependent ones"

**Comment by Zimmi48** [comment]
Wouldn't this kind of thing be doable in GitLab CI just by running `make install` for the dependency, then ensuring that `make ci-foo` for the next stage can find its dependencies as part of the already installed stuff? In fact, the present situation with `make ci-*` targets is not satisfying: they do not depend on `make world` because that would be detrimental on GitLab CI were Coq has been installed first. And this is annoying when doing local development. But they depend on each others and this is preventing us from just installing the dependency, going to another phase and building the next job. So shouldn't we keep these `make ci-*` targets for local development only (and make them depend on `world`) and use CI job dependency ordering instead for CI testing?

**Comment by ejgallego** [comment]
You could do that but latency would be pretty higher than what we have today. Keep in mind that stages have to _fully_ run before the next one fires up.

Regarding makefile targets I think that we won't be able to have a decent system because of lack of compositionality, you can hack around with the phony targets but that's no so useful IMHO.

In fact doing better on that respect was one of the main motivations for trying out Dune (and it works), there you can indeed drop coq and a contrib in the same directory and you get a global build rule for free.

**Comment by Zimmi48** [comment]
I kind of guessed that Dune would do it better. I hope we can try that soon.
With respect to the "stage must fully run" issue, I guess it then depends on whether we want a build that's faster overall or has shorter total build time (and thus takes a lighter load on our up-coming private runners).

**Comment by SkySkimmer** [comment]
The migration is basically done so maybe we should close this issue?

Some links related to more complex job topologies:

https://gitlab.com/gitlab-org/gitlab-ce/issues/29347
https://gitlab.com/gitlab-org/gitlab-ce/issues/35222
https://gitlab.com/gitlab-org/gitlab-ce/issues/41947
https://gitlab.com/gitlab-org/gitlab-ce/issues/47063

**Comment by ejgallego** [comment]
Indeed, congrats to all involved !! I was tho thinking of keeping this issue open until we receive our AWS/private runners setup, but we could as well open a new issue for this. Let's hear first what the consortium cc @maximedenes has to say.

**Comment by Zimmi48** [comment]
There's also the question of providing instructions for external project to setup CI using GitLab. But this definitely deserves a separate issue: cf. #7753.

**Comment by Zimmi48** [comment]
FTR I now have access to Coq's Inria servers and Pierre suggested to me to use the pyrolyse machine as a runner for our GitLab CI.

**Comment by maximedenes** [comment]
> FTR I now have access to Coq's Inria servers and Pierre suggested to me to use the pyrolyse machine as a runner for our GitLab CI.

Yes, I think it makes a lot of sense. Or we can keep it for benchmarks if pendulum is overloaded (but I don't think it is the case today).

**Comment by ejgallego** [comment]
Indeed we can start thinking of closing this issue, however OSX and Windows setup do remain to be documented, so we may want to wait a bit.

**Comment by ejgallego** [comment]
I think we can say "amazing work folks"

**Comment by maximedenes** [comment]
Yep, congrats! The CI infrastructure is really impressive these days.
